runtime/parser/lexer/lexer.go

```diff
@@ -49,6 +49,14 @@ type position struct {
 	column int
 }
 
+type LexerMode int
+
+const (
+	NORMAL = iota
+	STR_IDENTIFIER
+	STR_EXPRESSION
```

Comment:
Looks like `STR_EXPRESSION` is not set anywhere. So it actually on requires two modes? Maybe just use a boolean flag in that case. e.g: `inStringTemplate` or something similar.

Reply:
For this PR, `STR_EXPRESSION` could be removed and switched to a boolean flag instead, future extension for `${expression}` would require this to be added back potentially. With some limitations on syntax the changes for `${expression}` should not be too significant as in #3587.